test/performance/odp_dmafwd.c

```diff
@@ -0,0 +1,973 @@
+	pool_param.pkt.len = config->pkt_len;
+	pool_param.pkt.num = config->num_pkts;
+	pool_param.type = ODP_POOL_PACKET;
+	config->pktio_pool = odp_pool_create(PROG_NAME, &pool_param);
+
+	if (config->pktio_pool == ODP_POOL_INVALID) {
+		ODPH_ERR("Error creating packet I/O pool\n");
+		return false;
+	}
+
+	for (uint32_t i = 0U; i < config->num_ifs; ++i) {
+		pktio = &config->pktios[i];
+		odp_pktio_param_init(&pktio_param);
+		pktio_param.in_mode = ODP_PKTIN_MODE_SCHED;
+		pktio_param.out_mode = ODP_PKTOUT_MODE_DIRECT;
+		pktio->handle = odp_pktio_open(pktio->name, config->pktio_pool, &pktio_param);
+
+		if (pktio->handle == ODP_PKTIO_INVALID) {
+			ODPH_ERR("Error opening packet I/O (%s)\n", pktio->name);
+			return false;
+		}
+
+		odp_pktin_queue_param_init(&pktin_param);
+		pktin_param.queue_param.sched.prio = odp_schedule_default_prio();
+
+		if (odp_pktin_queue_config(pktio->handle, &pktin_param) < 0) {
+			ODPH_ERR("Error configuring packet I/O input queues (%s)\n", pktio->name);
+			return false;
+		}
+
+		odp_pktout_queue_param_init(&pktout_param);
+
+		if (odp_pktout_queue_config(pktio->handle, &pktout_param) < 0) {
+			ODPH_ERR("Error configuring packet I/O output queues (%s)\n", pktio->name);
+			return false;
+		}
+
+		if (odp_pktout_queue(pktio->handle, &pktio->out_q, 1) != 1) {
+			ODPH_ERR("Error querying packet I/O output queue (%s)\n", pktio->name);
+			return false;
+		}
+
+		if (odp_pktio_start(pktio->handle) < 0) {
+			ODPH_ERR("Error starting packet I/O (%s)\n", pktio->name);
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static void drain(void)
+{
+	odp_event_t ev;
+	odp_event_type_t type;
+	odp_dma_result_t res;
+	odp_buffer_t buf;
+	transfer_t *trs;
+
+	while (true) {
+		ev = odp_schedule(NULL, ODP_SCHED_NO_WAIT);
+
+		if (ev == ODP_EVENT_INVALID)
+			break;
+
+		type = odp_event_type(ev);
+
+		if (type == ODP_EVENT_DMA_COMPL) {
+			memset(&res, 0, sizeof(res));
+			odp_dma_compl_result(odp_dma_compl_from_event(ev), &res);
+			buf = (odp_buffer_t)res.user_ptr;
+			trs = (transfer_t *)odp_buffer_addr(buf);
+			odp_packet_free_multi(trs->src_pkts, trs->num);
+			odp_packet_free_multi(trs->dst_pkts, trs->num);
+			odp_buffer_free(buf);
+		}
+
+		odp_event_free(ev);
+	}
+}
+
+static int process_packets(void *args)
+{
+	thread_config_t *config = args;
+	odp_atomic_u32_t *is_running = &config->prog_config->is_running;
+	const int burst_size = config->prog_config->burst_size;
+	odp_event_t evs[burst_size];
+	int num_evs = 0;
+	uint8_t cnt = 0U;
+	const uint8_t num_ifs = config->prog_config->num_ifs;
+	pktio_t *pktios = config->prog_config->pktios;
+	copy_fn_t copy_fn = config->prog_config->copy_fn;
+
+	odp_barrier_wait(&config->prog_config->init_barrier);
+
+	while (odp_atomic_load_u32(is_running)) {
+		num_evs += odp_schedule_multi_no_wait(NULL, &evs[num_evs], burst_size - num_evs);
+
+		if (num_evs < burst_size)
+			continue;
+
+		copy_fn(evs, num_evs, &pktios[cnt++ % num_ifs], config);
```

Comment:
Changing the output interface like this can be a bit confusing to a user. Perhaps it's better to use static input-output queue pairs.

Reply:
Added a static lookup for this based on packet I/O index of received packets, similar to e.g. `odp_l2fwd` in v2. It's still not "perfect" though as there is potentially burst accumulation from several interfaces and only the packet I/O index of first to-be-sent-packet is checked.

Reply:
Refactored this further in v3. Packets are accumulated per packet I/O at reception and once required burst size is reached, copying is started.